Update cli to use StructOpt and add docs and tests

notmandatory opened the pull request:
#192 add docs for cli

notmandatory committed:
Update cli module to use StructOpt and add docs

notmandatory committed:
Fix docs

afilini commented on examples/repl.rs:
```diff
@@ -45,6 +43,10 @@ use bdk::sled;
 
 use bdk::blockchain::esplora::EsploraBlockchainConfig;
+use bdk::cli::{WalletOpt, WalletSubCommand};
+use bitcoin::hashes::core::str::FromStr;
```
I think you can also just import `std::str::FromStr` and it should work

notmandatory replied:
fixed

afilini commented on examples/repl.rs:
```diff
@@ -63,98 +65,97 @@ fn prepare_home_dir() -> PathBuf {
-    // let level = match matches.occurrences_of("v") {
-    //     0 => LevelFilter::Info,
-    //     1 => LevelFilter::Debug,
+    // let level = match cli_opt.verbosity.map(|v| v.len()) {
```
Is there any chance this is `None` here? I would map the info level to `None | Some(0)` just in case..

notmandatory replied:
I changed this around a bit to take a log level name, defaulting to "info". It may be possible to do the `-vv` style, but I didn't see an obvious way to do it with StructOpt, and using `-l error` style is a bit more explicit.

afilini commented on examples/repl.rs:
```diff
@@ -63,98 +65,97 @@ fn prepare_home_dir() -> PathBuf {
-        Some("regtest") => Network::Regtest,
-        Some("testnet") | _ => Network::Testnet,
-    };
+    let network = Network::from_str(cli_opt.network.as_str()).unwrap_or(Network::Testnet);
```
The `match` we were doing before here was mostly to disallow using mainnet for safety reason.. Maybe it still makes sense to do it?

Or ideally (that would be better than what was here before) we could either exit with an error or maybe print a warning that this is a development tool and it should be used carefully on mainnet

notmandatory replied:
I'm going to just a add a warning message if `bitcoin` is selected. There are certainly safe uses for `bdk` on mainnet such as using  monitoring / watchonly testing.

afilini replied:
Agreed, the quality has definitely improved since I originally wrote that code. A warning is probably more suitable given the current state of our code

notmandatory replied:
great, added a warning.

afilini commented on src/cli.rs:
```diff
@@ -22,10 +22,78 @@
+//! # use bdk::blockchain::{AnyBlockchain, ConfigurableBlockchain};
+//! # use bdk::blockchain::{AnyBlockchainConfig, ElectrumBlockchainConfig};
+//! # use bdk::cli;
+//! # use bdk::cli::{WalletOpt, WalletSubCommand};
```
You can also merge those two lines with `use bdk::cli::{self, WalletOpt, WalletSubCommand}`

notmandatory replied:
fixed.

afilini commented on examples/repl.rs:
```diff
@@ -63,98 +65,97 @@ fn prepare_home_dir() -> PathBuf {
+                            continue;
+                        }
+                        rl.add_history_entry(line.as_str());
+                        let mut repl_args: Vec<&str> = vec![" "];
```
I think you can avoid having to do this if you find a way to add the [`NoBinaryName`](https://docs.rs/clap/2.33.3/clap/enum.AppSettings.html#variant.NoBinaryName) flag. IIRC we were also doing this before somewhere here in the repl example.

That's useful when parsing something like a Vec/iterator of arguments.. basically whenever you are parsing a custom list of arguments rather than the command line from the OS. Not sure if it makes sense to have that as a default, but I would at least write somewhere in the docs that it can be done

notmandatory replied:
Unfortunately if I add the `NoBinaryName` setting to `WalletSubCommand` it works for repl but breaks when using it for cli args. As a work around I created a new `ReplOpt` struct in `repl.rs` with just a `WalletSubCommand` and setting `NoBinaryName` and this works. I also added some explaination in the docs.

notmandatory replied:
https://github.com/bitcoindevkit/bdk/pull/208/commits/364b47bfcb4a6f2ce4c9a6f34be92dbe4a301a73#diff-41438e9011cd34803a435657ff5534e6ff5a053c7b75b2d08fdcdbe7b309e78dR47

afilini reviewed (commented):
Great work, that looks much more organized than it was before!!

notmandatory commented:
> Great work, that looks much more organized than it was before!!

Thanks! and appreciate the detailed review. I addressed the issues you found, comments above.

notmandatory commented:
also have to thank @RCasatta for suggesting `StructOpt` it's a great tool for making the cli better documented and much more concise.

afilini commented:
ACK 364b47bfcb4a6f2ce4c9a6f34be92dbe4a301a73
